constrain the crypto protocols we use per android api level

crankycoder opened the pull request:
This should fix #1645

crankycoder committed:
added TLS patches to constrain the available crypto ciphers and
protocols that we use

crankycoder commented on libraries/stumbler/src/main/java/org/mozilla/mozstumbler/svclocator/services/log/DebugLogger.java:
```diff
@@ -28,7 +31,10 @@ public String e(String logTag, String s, Throwable e) {
             msg = "";
         } else {
-            msg = e.toString();
+            StringWriter sw = new StringWriter();
```
I reverted the DebugLogger to spew full stacktraces when exceptions are passed into the Log.e handler.

Note that this only affects _debug_ builds and not release builds.

crankycoder commented on libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/core/http/HttpUtil.java:
```diff
@@ -102,7 +96,8 @@ public String getUrlAsString(String url) throws IOException {
 
     @Override
     public InputStream getUrlAsStream(String url) throws IOException {
-        return new URL(url).openStream();
```
I had 2 different ways of handling URLs before - this has been consolidated so that we always use the mozOpenConnection function for URLs from now on.

crankycoder commented on libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/core/http/HttpUtil.java:
```diff
@@ -51,33 +52,26 @@ public HttpUtil(String ua) {
+                    null);
+            NoSSLv3SocketFactory factory = new NoSSLv3SocketFactory(sslcontext.getSocketFactory());
+            HttpsURLConnection.setDefaultSSLSocketFactory(factory);
+            return l_url.openConnection();
```
HTTPConnection object go through 2 phases - openConnection() and connect().  Between the two you can set options like timeouts on the connection.  Note that we don't call connect() here - some other bits of Android/Java do that for us.

crankycoder commented on libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/core/http/HttpUtil.java:
```diff
@@ -220,13 +213,13 @@ public IResponse post(String urlString, byte[] data, Map<String, String> headers
             httpURLConnection.setRequestMethod("POST");
-        } catch (IOException e) {
+        } catch (Exception e) {
             ClientLog.e(LOG_TAG, "Couldn't open a connection: ", e);
```
Crap.  This should be Log.

crankycoder replied:
Oh wait. I'm just tired and ClinetLog will just do the right thing if there is no GUI.

garvankeeley commented on android/build.gradle:
```diff
@@ -261,6 +261,7 @@ dependencies {
     // Stumbler library - always build from source
     compile project(':libraries:stumbler')
     testCompile project(':libraries:stumbler')
+
```
remove android.grade from commit

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/client/leaderboard/LBStumblerBundleReceiver.java:
```diff
@@ -51,6 +51,8 @@ public LBStumblerBundleReceiver(Context c) {
     @Override
     public void onReceive(Context context, Intent intent) {
+
+
```
remove LBStumblerBundleReceiver from commit

crankycoder commented on libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/core/http/GlobalConstants.java:
```diff
@@ -0,0 +1,67 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Add a link to Fennec code base

garvankeeley commented:
needs a rebase/merge

crankycoder committed:
added comment linking GlobalConstants back to original fennec source

crankycoder commented:
Don't merge this yet until I've verified correctness on my Nexus 4. I'm still seeing some odd behavior.

crankycoder committed:
force cipher suites to use GlobalDefaults and protocols to use GlobalDefaults

crankycoder committed:
oops.  this is the same bug as what we had in fennec for failed uploads with wifi only.

crankycoder commented:
@garvankeeley let's hold off on merging this until we can test on the old HTC gingerbread device in office.

crankycoder committed:
force a TLS POST when the debug ACRA option is triggered
added logging at level info if we're running in a debug build and someone does a POST

crankycoder commented:
Verified fixed on Android 2.3.6 

I/Stumbler_HttpUtil(  927): Negotiated HTTPS ciphers: TLS_RSA_WITH_AES_128_CBC_SHA
